opennms-services/src/main/java/org/opennms/netmgt/poller/monitors/DNSResolutionMonitor.java

```diff
@@ -28,115 +28,113 @@
-            } 
-            if (requireV4 && !v4found) {
-                String reason = nodeLabel + " could only be resolved to an IPv6 address";
-                LOG.debug(reason);
-                return PollStatus.unavailable(reason);
-            }
-            if (requireV6 && !v6found) {
-                String reason = nodeLabel + " could only be resolved to an IPv4 address";
-                LOG.debug(reason);
-                return PollStatus.unavailable(reason);
+        // Build a resolver object used for lookups
+        final String nameserver = ParameterMap.getKeyedString(parameters,
+                                                              PARM_NAMESERVER,
+                                                              null);
+
+        final Resolver resolver;
+        try {
+            if (nameserver == null) {
+                // Use system-defined resolvers
+                resolver = new ExtendedResolver();
+            } else {
+                resolver = new SimpleResolver(nameserver);
             }
-            LOG.debug("Resolved {} correctly!", nodeLabel);
-            return PollStatus.available((double)(end - start));
 
-        } catch (TextParseException e) {
-            String reason = "Unable to resolve "+nodeLabel+": "+e.getMessage();
-            LOG.debug(reason);
-            return PollStatus.unavailable(reason);
+        } catch (final UnknownHostException e) {
+            return PollStatus.unavailable("Unable to resolve nameserver '" + nameserver + "': " + e.getMessage());
         }
 
-    }
-    
-    InetAddress[] resolve(String hostname) throws TextParseException {
-        Record[] aaaaRecords = new Lookup(hostname, Type.AAAA).run();
-        Record[] aRecords = new Lookup(hostname, Type.A).run();
-        
-        InetAddress[] addrs = new InetAddress[(aaaaRecords == null ? 0 : aaaaRecords.length)+(aRecords == null ? 0 : aRecords.length)];
-        
-        int index = 0;
-        if (aaaaRecords != null) {
-            for(Record r : aaaaRecords) {
-                addrs[index++] = ((AAAARecord)r).getAddress();
-            }
-        }
-        
-        if (aRecords != null) {
-            for(Record r : aRecords) {
-                addrs[index++] = ((ARecord)r).getAddress();
-            }
+        // Start resolving the records
+        final long start = System.currentTimeMillis();
+
+        // Resolve the name
+        final boolean ipv4Found = resolve(name, resolver, Type.A);
+        final boolean ipv6Found = resolve(name, resolver, Type.AAAA);
+
+        // Resolving succeeded - checking results
+        final long end = System.currentTimeMillis();
+
+        // Check if result is valid
+        if (!ipv4Found && !ipv6Found) {
+            return PollStatus.unavailable("Unable to resolve host '" + name + "'");
+
+        } else  if (ipv4Required && !ipv4Found) {
+            return PollStatus.unavailable("No IPv4 Address result (A record) returned for '" + name + "'");
```

Comment:
In the old monitor, this was:
```java
name + " could only be resolved to an IPv6 address"
```
indicating that an AAAA record was found but no A record was found. This is not apparent in the new message. Can we make the messages for this something like:
```java
name + " could be resolved to an IPv6 address (AAAA record) but not an IPv4 address (A record)."
```